src/si/helpers.stanza

```diff
@@ -88,6 +89,143 @@ public defn topo-net (pair:KeyValue<JITXObject,JITXObject|KeyValue>):
+  matches port `d`, and so on.
+
+  Returns a tuple of all newly-instantiated objects.
+
+  @snippet
+
+  ```stanza
+  port mii : rmii-bus
+  require sw-bus : rmii-bus from mcu
+  val series-R-type = create-resistor(resistance = 51.0, precision = (1 %), case = ["0603"])
+  val rx-bus = [sw-bus.rxd, sw-bus.crs-dv, sw-bus.ref-clk]
+  val rx-out = [mii.rxd, mii.crs-dv, mii.ref-clk]
+  topo-bus(ShortTrace(rx-bus => series-R-type) => rx-out)
+  ```
+
+  @snip-note 3 `create-resistor` returns an `Instantiable`
+  @snip-note 4 `sw-bus.rxd` and `mii.rxd` are both arrays of two pins. `topo-bus` will return a tuple of four new resistors that connect the two buses.
+  @snip-note 6 All objects in `rx-bus` will be connected to their respective resistor with a short-trace.
+<DOC>
+public defn topo-bus (pair:KeyValue) -> Tuple<JITXObject> :
+  ensure-valid-tree!(pair)
+  val [l, r] = unpack(pair)
+  val acc = Vector<JITXObject>()
+  make-topo-bus(l, r, acc, short-trace? = pair is ShortTrace)
+  to-tuple $ acc
+
+defn ensure-valid-tree! (kv:KeyValue) : false ; [TODO] : check precondition
+
+; Traverse left, then connect to the right
+defn make-topo-bus (l:KeyValue|JITXObject|Tuple, r:KeyValue|JITXObject|Tuple|Instantiable, acc:Vector, -- short-trace?:True|False = false) -> Tuple<JITXObject> :
+  val lhs = flatten(l, acc)
+  connect(lhs, r, acc, short-trace?)
+
+defn connect (lhs:Tuple, r:KeyValue|JITXObject|Tuple|Instantiable, acc:Vector, short-trace?) -> Tuple :
+  match(r) :
+    (x:KeyValue) :
+      val [l, r] = unpack(x)
+      val lhs* = connect(lhs, l, short-trace?, acc)
+      connect(lhs*, r, acc, x is ShortTrace)
+    (x:JITXObject) :
+      connect-topo-bus(lhs, flatten-obj(x), short-trace?)
+    (x:Tuple) :
+      val r* = to-tuple $ seq-cat(flatten-obj, x)
+      connect-topo-bus(lhs, r*, short-trace?)
+    (x:Instantiable) :
+      connect-topo-bus(lhs, x, acc, short-trace?)
+
+defn connect-topo-bus (lhs:Tuple, rhs:Instantiable, acc, short-trace?:True|False) -> Tuple :
+  inside pcb-module :
+    to-tuple $
+      for l in lhs seq :
+        inst i : rhs 
+        add(acc, i)
+        topo-net(l, i.p[1])
+        if short-trace? :
+          short-trace(l, i.p[1])
+        i.p[2]
+
+defn connect-topo-bus (lhs:Tuple, rhs:Tuple, short-trace?:True|False) -> Tuple :
+  inside pcb-module :
+    to-tuple $
+      for (a in lhs, b in rhs) seq :
+        topo-net(a => b)
+        if short-trace? :
+          short-trace(a, b)
+        b
+
+defn flatten (x:KeyValue|JITXObject|Tuple, acc:Vector) -> Tuple<JITXObject> :
+  match(x) :
+    (x:KeyValue) :
+      val [l, r] = unpack(x)
+      make-topo-bus(l, r, acc, short-trace? = x is ShortTrace)
+    (x:JITXObject) : flatten-obj(x)
+    (x:Tuple) : to-tuple $ seq-cat(flatten-obj, x)
+
+defn flatten-obj (x:JITXObject) -> Tuple<JITXObject> :
+  match(port-type(x)) :
+    (t:SinglePin) : [x]
+    (t:PortArray) : to-tuple $ pins(x)
+    (otherwise) : throw $ Exception("nope")
```

Comment:
I've been using `ValueError` here to have something consistent.

Reply:
Maybe a `TypeError` would be more reasonable here.

Reply:
yeah -- fixing all the exceptions today, they're all useless right now